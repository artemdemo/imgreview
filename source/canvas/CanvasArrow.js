import Konva from 'konva';
import _throttle from 'lodash/throttle';

class CanvasArrow {
    constructor(mainStage) {
        this._stage = mainStage;
        this._curveLayer = null;
        this._anchorLayer = null;
        this._quad = null;
    }

    drawArrow = () => {
        const context = this._curveLayer.getContext();

        context.clear();

        // draw quad
        context.beginPath();
        context.moveTo(
            this._quad.start.attrs.x,
            this._quad.start.attrs.y,
        );
        context.quadraticCurveTo(
            this._quad.control.attrs.x,
            this._quad.control.attrs.y,
            this._quad.end.attrs.x,
            this._quad.end.attrs.y,
        );
        context.setAttr('strokeStyle', 'red');
        context.setAttr('lineWidth', 4);
        context.stroke();
    };

    buildAnchor(x, y) {
        const anchor = new Konva.Circle({
            x,
            y,
            radius: 20,
            stroke: '#666',
            fill: '#ddd',
            strokeWidth: 2,
            draggable: true,
        });

        const self = this;

        // add hover styling
        anchor.on('mouseover', function() {
            document.body.style.cursor = 'pointer';
            this.setStrokeWidth(4);
            self._anchorLayer.draw();
        });
        anchor.on('mouseout', function() {
            document.body.style.cursor = 'default';
            this.setStrokeWidth(2);
            self._anchorLayer.draw();
        });

        anchor.on('dragend', this.drawArrow);

        anchor.on('dragmove', _throttle(this.drawArrow, 50));

        this._anchorLayer.add(anchor);
        return anchor;
    }

    addArrow() {
        this._anchorLayer = new Konva.Layer();
        this._curveLayer = new Konva.Layer();

        this._quad = {
            start: this.buildAnchor(60, 30),
            control: this.buildAnchor(240, 110),
            end: this.buildAnchor(80, 160),
        };

        this._stage.add(this._curveLayer);
        this._stage.add(this._anchorLayer);

        this.drawArrow();
    }
}

export default CanvasArrow;
