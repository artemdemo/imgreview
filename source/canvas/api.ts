/**
 * This is public API to manipulate with canvas components.
 * Canvas should be isolated from the interface components.
 */

import emitter from './events/eventsEmitter';
import * as keys from './events/eventsKeys';
import { TCreateArrowOptions } from './events/eventsTypes';

export type TImageData = {
    image: any;
    name: string;
};

export type TCanvasSize = {
    width: number;
    height: number;
};

// Emitting events
//

export const createArrow = (options?: TCreateArrowOptions) => {
    emitter.emit(keys.CREATE_ARROW, options);
};

export const setImage = (data: TImageData) => {
    emitter.emit(keys.SET_IMAGE, data);
};

export const exportCanvasToImage = (name: string) => {
    emitter.emit(keys.EXPORT_CANVAS_TO_IMAGE, name);
};

export const updateCanvasSize = (data: TCanvasSize) => {
    emitter.emit(keys.UPDATE_CANVAS_SIZE, data);
};

export const blurShapes = () => {
    emitter.emit(keys.BLUR_SHAPES);
};

// Subscribing to events
//

export const onImageUpdate = (cb: () => void) => {
    emitter.on(keys.ON_IMAGE_UPDATE, cb);
};
