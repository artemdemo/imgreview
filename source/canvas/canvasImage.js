import Konva from 'konva';

class Image {
    constructor(mainStage) {
        this._stage = mainStage;
        this._layer = null;
        this._img = null;
    }

    /**
     * Set image to the stage
     * @param props {object} https://konvajs.github.io/api/Konva.Image.html
     */
    setImage(props) {
        this._layer = new Konva.Layer();
        this._img = new Konva.Image(props);
        this._layer.add(this._img);
        this._stage.add(this._layer);
    }
}

export default Image;
