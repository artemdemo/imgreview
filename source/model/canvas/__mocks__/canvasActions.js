export const setStage = jest.fn();
export const addImage = jest.fn();
export const updateImageSize = jest.fn();
export const saveCanvas = jest.fn();
export const setCursor = jest.fn();
