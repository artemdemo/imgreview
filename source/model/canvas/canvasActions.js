import { createAction } from 'redux-act';

export const setStage = createAction('SET_STAGE');
export const addImage = createAction('ADD_IMAGE');
export const updateImageSize = createAction('UPDATE_IMAGE');
export const saveCanvas = createAction('SAVE_CANVAS');
export const setCursor = createAction('SET_CURSOR');
