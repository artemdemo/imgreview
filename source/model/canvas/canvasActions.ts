import { createAction } from 'redux-actions';

export const addImage = createAction('ADD_IMAGE');
export const setCursor = createAction('SET_CURSOR');

// ToDo: I don't like that I'm defining action type as a different exported type in the same file
// I'm doing it, because of `type Props = {}` that I'll need to define later in `MIResize.jsx` and `MIResizePopup.jsx`
// Later action will be injected via props and I want to get access to it's signature.
// Probably there should be an solution to that, but currently I don't know what's that.
export type TUpdateImageSize = (prop: { width: number, height: number }) => void;
export const updateImageSize = createAction('UPDATE_IMAGE');
