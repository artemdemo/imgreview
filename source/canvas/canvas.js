import Konva from 'konva';
import CanvasImage from './CanvasImage';
import CanvasArrow from './CanvasArrow';
import { drawImageScaled, drawImageAsIs } from './drawImage';

let stage;
let container;

export const init = (containerEl) => {
    container = containerEl;
};

export const setStage = (width, height) => {
    stage = new Konva.Stage({
        container,
        width,
        height,
    });
};

export const drawImage = (image) => {
    setStage(image.width, image.height);

    const imgRef = new CanvasImage(stage);
    imgRef.setImage({
        image,
    });

    // _context.clearRect(0, 0, _canvas.width, _canvas.height);
    // if (img.width > _canvas.width || img.height > _canvas.height) {
    //     drawImageScaled(img, _canvas, _context);
    // } else {
    //     drawImageAsIs(img, _canvas, _context);
    // }
};

export const addArrow = () => {
    const arrowRef = new CanvasArrow(stage);
    arrowRef.addArrow();
};
