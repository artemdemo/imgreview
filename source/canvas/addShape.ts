/* eslint-disable import/prefer-default-export */

import _get from 'lodash/get';
import store from '../store';
import canvasStore from './store';
import { cursorTypes } from '../model/canvas/canvasConst';
import { blurShapes, addArrow } from '../model/shapes/shapesActions';
import { setImage } from './model/image/imageActions';
import { setCursor } from '../model/canvas/canvasActions';
import CanvasImage from './Image/CanvasImage';
import Arrow from './Arrow/Arrow';
import { TImageData } from './api';

/**
 * Connect Arrow to the stage.
 * If arrow provided - it will use provided instance,
 * if not - will create new one.
 * @param arrow {Arrow|null} - I'm using it when coping Arrows.
 * @param options {object}
 * @param options.strokeColor {string}
 * @param options.strokeWidth {number}
 */
export const connectArrow = (arrow?: Arrow|null, options?: { strokeColor: string, strokeWidth: number }) => {
    const { stage } = <any> canvasStore.getState();
    const _arrow = arrow || new Arrow({
        stroke: _get(options, 'strokeColor', '#000'),
        strokeWidth: _get(options, 'strokeWidth', 5),
    });
    _arrow.addToStage(stage.instance);
    _arrow.on('click', arrowInstance => store.dispatch(blurShapes(arrowInstance)));
    _arrow.on('dragstart', arrowInstance => store.dispatch(blurShapes(arrowInstance)));
    _arrow.on('mouseover', () => store.dispatch(setCursor(cursorTypes.move)));
    _arrow.on('mouseout', () => store.dispatch(setCursor(cursorTypes.auto)));
    _arrow.onAnchor('mouseover', () => store.dispatch(setCursor(cursorTypes.pointer)));
    _arrow.onAnchor('mouseout', () => store.dispatch(setCursor(cursorTypes.auto)));
    store.dispatch(addArrow(_arrow));
};


export const addImageToStage = (data: TImageData) => {
    const { stage, image } = <any> canvasStore.getState();
    if (image.instance) {
        image.instance.destroy();
    }
    stage.instance.setAttr('width', data.image.width);
    stage.instance.setAttr('height', data.image.height);
    const canvasImage = new CanvasImage({
        image: data.image,
    });
    canvasImage.addToStage(stage.instance);
    canvasStore.dispatch(setImage({
        image: canvasImage,
    }));
};
