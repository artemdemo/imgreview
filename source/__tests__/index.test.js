import React from 'react';
import renderer from 'react-test-renderer';
import { updateImageSize } from '../model/canvas/canvasActions';
import { App, onImageUpdateCb } from '../index';

jest.mock('react-redux');
jest.mock('react-dom');
jest.mock('../store');
jest.mock('../../srcCanvas/api');
jest.mock('../services/document');
jest.mock('../views/AppView.async');

const reactDomMock = require('react-dom');
const documentMock = require('../services/document');
const apiMock = require('../../srcCanvas/api');
const storeMock = require('../store');

describe('index', () => {
    it('should render App', () => {
        const tree = renderer.create(
            <App />
        ).toJSON();

        expect(tree).toMatchSnapshot();
        expect(documentMock.getElementById).toBeCalledWith('app');
        expect(reactDomMock.render).toBeCalledWith(
            <App />,
            undefined,
        );
    });

    it('should set cb onImageUpdate', () => {
        expect(apiMock.onImageUpdate).toBeCalledWith(onImageUpdateCb)
    });

    it('onImageUpdate callback should dispatch action', () => {
        const data = {
            content: 'Some data',
        };
        onImageUpdateCb(data);
        expect(storeMock.default.dispatch).toBeCalledWith(updateImageSize(data));
    });
});
