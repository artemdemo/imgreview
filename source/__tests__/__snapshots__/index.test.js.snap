// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index should render App 1`] = `
<div
  data-mock="ReactReduxProvider"
>
  <div
    data-mock="AppViewAsync"
  />
</div>
`;
